source/Octopus.Tentacle/Kubernetes/Diagnostics/KubernetesAgentMetrics.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Util;
+
+namespace Octopus.Tentacle.Kubernetes.Diagnostics
+{
+    public class KubernetesAgentMetrics
+    {
+        readonly string lastEventTimestampKey = "latestTimestamp";
+
+        public delegate KubernetesAgentMetrics Factory(IPersistenceProvider persistenceProvider);
+
+        readonly IPersistenceProvider persistenceProvider;
+        readonly ISystemLog log;
+
+        public KubernetesAgentMetrics(IPersistenceProvider persistenceProvider, ISystemLog log)
+        {
+            this.persistenceProvider = persistenceProvider;
+            this.log = log;
+        }
+
+        public void TrackEvent(string reason, string source, DateTimeOffset occurrence)
+        {
+            DateTimeOffset latestEvent;
+            try
+            {
+                latestEvent = GetLatestEventTimestamp();
```

Comment:
I'm not a huge fan of public methods depending on other public methods in general.
In this case if you made a separate `GetLatestEventTimestampInternal` method that has all the code of the `GetLatestEventTimestamp` method except for the lock, then you can call the internal one directly in here and only need one lock to wrap the entire `TrackEvent` method. This saves having to unlock and then relock.

Reply:
Not an idea I'd come across, but I kinda like it.